crates/cubecl-core/src/ir/local_allocator.rs

```diff
@@ -0,0 +1,117 @@
+
+type ScopeRef = Rc<RefCell<Scope>>;
+
+/// Defines a local variable allocation strategy (i.e. reused mutable variables, SSA)
+pub trait LocalAllocator {
+    /// Creates a local variable that can be (re)assigned
+    fn create_local_variable(&self, root: ScopeRef, scope: ScopeRef, item: Item) -> ExpandElement;
+    /// Creates an immutable local binding for intermediates
+    fn create_local_binding(&self, root: ScopeRef, scope: ScopeRef, item: Item) -> ExpandElement;
+}
+
+#[derive(Default, Clone)]
+pub struct VariablePool {
+    map: Rc<RefCell<HashMap<Item, Vec<ExpandElement>>>>,
+}
+
+impl VariablePool {
+    /// Returns an old, not used anymore variable, if there exists one.
+    pub fn reuse(&self, item: Item) -> Option<ExpandElement> {
+        let map = self.map.borrow();
+
+        // Filter for candidate variables of the same Item
+        let variables = map.get(&item)?;
+
+        // Among the candidates, take a variable if it's only referenced by the map
+        // Arbitrarily takes the first it finds in reverse order.
+        for variable in variables.iter().rev() {
+            match variable {
+                ExpandElement::Managed(var) => {
+                    if Rc::strong_count(var) == 1 {
+                        return Some(variable.clone());
+                    }
+                }
+                ExpandElement::Plain(_) => (),
+            }
+        }
+
+        // If no candidate was found, a new var will be needed
+        None
+    }
+
+    /// Insert a new variable in the map, which is classified by Item
+    pub fn insert(&self, var: ExpandElement) {
+        let mut map = self.map.borrow_mut();
+        let item = var.item();
+
+        if let Some(variables) = map.get_mut(&item) {
+            variables.push(var.clone());
+        } else {
+            map.insert(var.item(), vec![var.clone()]);
+        }
+    }
+}
+
+/// Reusing allocator, assigns all intermediates to a set of mutable variables that get continuously
+/// reused.
+#[derive(Default)]
+pub struct ReusingAllocator {
+    pool: VariablePool,
+}
+
+impl LocalAllocator for ReusingAllocator {
+    fn create_local_variable(&self, root: ScopeRef, scope: ScopeRef, item: Item) -> ExpandElement {
+        if item.elem.is_atomic() {
+            let new = scope.borrow_mut().create_local_undeclared(item);
+            return ExpandElement::Plain(new);
+        }
+
+        // Reuse an old variable if possible
+        if let Some(var) = self.pool.reuse(item) {
+            return var;
+        }
+
+        // Create a new variable at the root scope
+        // Insert it in the variable pool for potential reuse
+        let new = ExpandElement::Managed(Rc::new(root.borrow_mut().create_local(item)));
+        self.pool.insert(new.clone());
+
+        new
+    }
+
+    fn create_local_binding(&self, root: ScopeRef, scope: ScopeRef, item: Item) -> ExpandElement {
+        self.create_local_variable(root, scope, item)
+    }
```

Comment:
I'm not sure I undetstand the fundamental difference between `local_variable` and `local_binding`.

Reply:
What is the logic behind calling `local_binding` vs `local_variable` ? When do you know you should do one vs the other?

Reply:
Local bindings are things that are only assigned to once. So intermediates, immutable variables, etc. The term that's mostly used these days is "binding", which is why I called it that. It would be `let x = ` in `wgsl`, as opposed to "variables", which are a `var x` in `wgsl` and can be reassigned. In the generated SPIR-V, `binding`s would become normal SSA ops (i.e. `%32 = AddOp %30 %31`), while `variable`s would become a variable declaration, along with a `LoadOp` for accessing it and a `StoreOp` for writing to it. An `AddOp` with a `variable` would result in code like this
```rust
%30 = LoadOp %3
%31 = LoadOp %5
%32 = AddOp %30 %31
StoreOp %32 %3
```

Reply:
The important difference is that bindings don't necessarily exist after compilation, they're just labels for intermediate values. Variables do exist at runtime and must be used with load/store.

Reply:
As for when you'd use either: you'd use a local binding by default, and only use a variable when you need to reassign the value (i.e. `let mut` or `vectorized_empty`).